### PR TITLE
Add Reaper Control Plane Deployment Mode

### LINKED ISSUE #1277: New Reaper deployment mode: Control Plane
We want to allow deploying a single instance of Reaper which will repair all K8ssandraClusters managed by a single control plane.
This requires to create a new deployment mode in Reaper (aside from PER_DC and SINGLE), which we can name "CONTROL_PLANE".
A Reaper custom resource would need to be created separately prior to the K8ssandraCluster objects creation, with this mode.
Any K8ssandraCluster object without Reaper settings would be registered by the operator into the control plane Reaper and repairs would be managed by it.

We would disallow enabling Reaper in clusters if the control plane instance exists to keep things simple in the first iteration.
This means either you have one centralized Reaper managing all clusters or you have individual Reaper instances per cluster.

```[tasklist]
### Definition of Done
- [x] the Reaper API has a new deployment mode named `CONTROL_PLANE`
- [x] A Reaper running with memory storage, the depoyment mode must be `CONTROL_PLANE` or `SINGLE`
- [x] A `CONTROL_PLANE` reaper comes up without contacting any K8ssandra clusters (for storage or registration)
- [x] A `SINGLE` reaper won't deploy more than once
- [x] The k8ssandraCluster object has a field in the reaper spec that defines which Reaper to register with
- [ ] ~In the K8ssandraCluster spec, a reaperRef is added to point to the Reaper instance that will handle the repairs~
- [x] When a CP Reaper instance exists, k8ssandra-operator registers all K8ssandraClusters in it
- [ ] ~If a K8ssandraCluster object gets created with Reaper settings while a control plane Reaper exists, the K8ssandraCluster objects should override control plane Reaper with their own defined Reaper settings.~
```

### PR DESCRIPTION
**Pull request by rzvoncek**
**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #1277

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go [hunk 1]
```diff
@@ -280,3 +290,30 @@ func (r *K8ssandraCluster) ValidateMedusa() error {
...
+			return ErrNoReaperAccessMode
+		}
+		if r.Spec.Reaper.StorageConfig.Resources.Requests == nil {
+			return ErrNoReaperResourceRequests
+		}
+		if r.Spec.Reaper.StorageConfig.Resources.Requests.Storage().IsZero() {
+			return ErrNoReaperStorageRequest
+		}
+	}
+	if r.Spec.Reaper.StorageType == reaperapi.StorageTypeLocal && r.Spec.Reaper.DeploymentMode == reaperapi.DeploymentModePerDc {
```
Comment (rzvoncek) [inline apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go:315]: I'm not sure if we need to enforce this here. Reaper won't come up, in this situation, so it's not the end of the world. On the other hand, it broke one e2e test, so perhaps it'll break stuff for users out there also.

File: controllers/reaper/reaper_controller.go [hunk 2]
```diff
@@ -248,6 +298,29 @@ func (r *ReaperReconciler) reconcileDeployment(
...
 }
 
+func getDeploymentReplicas(actualDeployment client.Object, logger logr.Logger) int32 {
+	switch actual := actualDeployment.(type) {
+	case *appsv1.Deployment:
+		return *actual.Spec.Replicas
+	case *appsv1.StatefulSet:
+		return *actual.Spec.Replicas
+	default:
+		logger.Error(fmt.Errorf("unexpected type %T", actualDeployment), "Failed to get deployment replicas")
```
Comment (adejanovski) [inline controllers/reaper/reaper_controller.go:308]: **Suggestion:** we should probably fail the reconcile here instead of swallowing the error. The return type should then be `(int32, error)` and if an error occurs it should be returned.

File: controllers/reaper/reaper_controller.go [hunk 3]
```diff
@@ -248,6 +298,29 @@ func (r *ReaperReconciler) reconcileDeployment(
...
+}
+
+func forceSingleReplica(desiredDeployment *client.Object, logger logr.Logger) {
+	switch desired := (*desiredDeployment).(type) {
+	case *appsv1.Deployment:
+		desired.Spec.Replicas = ptr.To[int32](1)
+	case *appsv1.StatefulSet:
+		desired.Spec.Replicas = ptr.To[int32](1)
+	default:
+		logger.Error(fmt.Errorf("unexpected type %T", desiredDeployment), "Failed to set deployment replicas")
```
Comment (adejanovski) [inline controllers/reaper/reaper_controller.go:320]: **Suggestion:** same here, an error should be returned.

**Review by adejanovski — CHANGES_REQUESTED** [review]
This is coming to shape nicely!
I have a few suggestions that should hopefully be easy to implement in order to harmonize with how we do things elsewhere in the operator.
Could you create a ticket to update the docs as well, we'll need to make it easy for users to use this new feature.
Thanks!

File: controllers/k8ssandra/reaper.go [hunk 4]
```diff
@@ -253,3 +264,27 @@ func getSingleReaperDcName(kc *api.K8ssandraCluster) string {
...
+	ctx context.Context,
+	kc *api.K8ssandraCluster,
+	actualDc *cassdcapi.CassandraDatacenter,
+	logger logr.Logger,
+) result.ReconcileResult {
+	manager := reaper.NewManager()
+	manager.SetK8sClient(r)
+	if username, password, err := manager.GetUiCredentials(ctx, kc.Spec.Reaper.UiUserSecretRef, kc.Namespace); err != nil {
+		logger.Error(err, "Failed to get Reaper UI user secret")
+		return result.RequeueSoon(r.DefaultDelay)
```
[the remark this answers is not in the document]
Comment (burmanm) [reply controllers/k8ssandra/reaper.go:278]: And it's not just resurfacing the error, but also backoff timer for the requeue. So if the error keeps persisting, the operator will increase the delay for requeue to avoid unnecessary load.

File: controllers/reaper/reaper_controller.go (hunk 2 above)
Comment (rzvoncek) [reply controllers/reaper/reaper_controller.go]: Pushed a commit solving this.

File: controllers/reaper/reaper_controller.go (hunk 3 above)
Comment (rzvoncek) [reply controllers/reaper/reaper_controller.go]: Pushed a commit solving this.

File: controllers/k8ssandra/reaper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/reaper.go]: Pushed a commit fixing this.

File: controllers/k8ssandra/reaper.go [hunk 5]
```diff
@@ -253,3 +264,27 @@ func getSingleReaperDcName(kc *api.K8ssandraCluster) string {
...
+	} else {
+		if err = manager.ConnectWithReaperRef(ctx, kc, username, password); err != nil {
+			logger.Error(err, "Failed to connect to external Reaper")
+			return result.RequeueSoon(r.DefaultDelay)
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/reaper.go]: Pushed a commit fixing this.

File: controllers/k8ssandra/reaper.go [hunk 6]
```diff
@@ -253,3 +264,27 @@ func getSingleReaperDcName(kc *api.K8ssandraCluster) string {
...
+		}
+		if err = manager.AddClusterToReaper(ctx, actualDc); err != nil {
+			logger.Error(err, "Failed to add cluster to external Reaper")
+			return result.RequeueSoon(r.DefaultDelay)
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/reaper.go]: Pushed a commit fixing this.

File: controllers/k8ssandra/secrets.go [hunk 7]
```diff
@@ -60,9 +64,12 @@ func (r *K8ssandraClusterReconciler) reconcileReaperSecrets(ctx context.Context,
 				uiUserSecretRef.Name = reaper.DefaultUiSecretName(kc.SanitizedName())
 			}
 			kcKey := utils.GetKey(kc)
-			if err := secret.ReconcileSecret(ctx, r.Client, cassandraUserSecretRef.Name, kcKey); err != nil {
-				logger.Error(err, "Failed to reconcile Reaper CQL user secret", "ReaperCassandraUserSecretRef", cassandraUserSecretRef)
-				return result.Error(err)
+			if kc.Spec.Reaper != nil && kc.Spec.Reaper.ReaperRef.Name == "" {
+				// ReaperRef presence indicates we use an Control Plane Reaper, which talks to C* via mgmt API, so we don't need a CQL user
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/secrets.go]: I agree. Checking for storage type is much cleaner. Pushed a commit doing this.

File: CHANGELOG/CHANGELOG-1.18.md [hunk 8]
```diff
@@ -23,3 +23,5 @@ When cutting a new release, update the `unreleased` heading to the tag being gen
 * [ENHANCEMENT] [#1274](https://github.com/k8ssandra/k8ssandra-operator/issues/1274) On upgrade, do not modify the CassandraDatacenter object unless instructed with an annotation `k8ssandra.io/autoupdate-spec` with value `once` or `always`
 * [BUGFIX] [#1222](https://github.com/k8ssandra/k8ssandra-operator/issues/1222) Consider DC-level config when validating numToken updates in webhook
 * [BUGFIX] [#1366](https://github.com/k8ssandra/k8ssandra-operator/issues/1366) Reaper deployment can't be created on OpenShift due to missing RBAC rule
+* [CHANGE] Update cassandra-medusa to 0.22.0
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply CHANGELOG/CHANGELOG-1.18.md]: I've rebased this on master after the other PR was merged. This PR no longer attempts to change the 1-18 changelog.

File: apis/reaper/v1alpha1/reaper_types.go [hunk 9]
```diff
@@ -218,17 +236,45 @@ type ReaperClusterTemplate struct {
...
+}
+
+func (rct *ReaperClusterTemplate) IsControlPlane() bool {
+	return rct != nil && rct.DeploymentMode == DeploymentModeControlPlane
+}
+
+// EnsureDeploymentMode ensures that a deployment mode is no PER_DC if we use the local storage type. This is to prevent
+// several instances of Reapers with local storage that would interfere with each other. Deployment mode CONTROL_PLANE
+// is allowed.
+func (t *ReaperClusterTemplate) EnsureDeploymentMode() bool {
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply apis/reaper/v1alpha1/reaper_types.go]: I've removed the check altogether and am returning an error instead.

File: apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go (hunk 1 above)
Comment (rzvoncek) [reply apis/k8ssandra/v1alpha1/k8ssandracluster_webhook.go:315]: I've actually added a local storage mode to one of the fixtures in the StatefulSet PR [here](https://github.com/k8ssandra/k8ssandra-operator/pull/1365/files#diff-39d6b74cc9e0e10063db88b693f835c51d8f28e56cf1e8d3b9628d0496c520f7).  This is used by the `CreateReaperAndDatacenter` e2e test.

File: controllers/k8ssandra/secrets.go [hunk 10]
```diff
@@ -41,6 +41,10 @@ func SuperuserSecretName(kc *api.K8ssandraCluster) string {
 }
 
 func (r *K8ssandraClusterReconciler) reconcileReaperSecrets(ctx context.Context, kc *api.K8ssandraCluster, logger logr.Logger) result.ReconcileResult {
+	if kc.Spec.Reaper != nil && kc.Spec.Reaper.ReaperRef.Name != "" {
+		logger.Info("ReaperRef points to an existing Reaper, we don't need a new Reaper CQL secret")
```
[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/secrets.go:46]: I don't think it's as straightforward here. But perhaps I still don't understand things well enough.

We are here in the k8ssandra cluster reconciliation loop, and we're checking what the kc.reaper.Spec tells us.
We are deciding what to do with the CQL secret and the UI secret. We're not dealing with the JMX (except if the CQL and JMX secrets are one and the same).

We only want to reconcile(~create) the secrets if we're using a k8ssandra-cluster-specific Reaper.

We can have a k8ssandra-cluster-specific Reaper with local storage. In that case we don't need a CQL secret, only the UI secret. This is checked on L68.

I think that's why we cannot check for just the storage on L44. I'd really need the `isControlPlane()` function somewhere, but I cannot have that for other reasons. The ReaperRef presence is still my favourite alternative to that.

To sum up, I think I'd need clarification of:
* how the JMX secret comes into this?
* do we actually want to allow _bundled_ Reapers with local storage?

[the remark this answers is not in the document]
Comment (adejanovski) [reply controllers/k8ssandra/secrets.go:46]: > do we actually want to allow bundled Reapers with local storage?

Yes, why not? It's an efficient storage mechanism which doesn't require using Cassandra itself.

> how the JMX secret comes into this?

The cases where we need Reaper to create users is for:

- authenticated access with JMX enabled
- using cassandra as storage for Reaper

One could have a centralized Reaper using JMX to communicate with the Cassandra clusters. It's not the default, but it's a possibility.
In this case, we still need a user and its secret to be created.

It is just a suggestion though and if that feels like some complications we want to deal with later, I'm good with that.

[the remark this answers is not in the document]
Comment (rzvoncek) [reply controllers/k8ssandra/secrets.go:46]: >> do we actually want to allow bundled Reapers with local storage?

> Yes, why not? It's an efficient storage mechanism which doesn't require using Cassandra itself.

I agree with this, so that's sorted. 

About the JMX, I still don't get why we worry about it here when the `reconcileReaperSecrets` does not mention JMX secrets at all.

I'm indeed leaning towards not accepting this suggestion and keeping the condition as it is.

[the remark this answers is not in the document]
Comment (adejanovski) [reply controllers/k8ssandra/secrets.go:46]: ok then 👍

File: controllers/k8ssandra/secrets.go [hunk 11]
```diff
@@ -60,9 +63,12 @@ func (r *K8ssandraClusterReconciler) reconcileReaperSecrets(ctx context.Context,
 				uiUserSecretRef.Name = reaper.DefaultUiSecretName(kc.SanitizedName())
 			}
 			kcKey := utils.GetKey(kc)
-			if err := secret.ReconcileSecret(ctx, r.Client, cassandraUserSecretRef.Name, kcKey); err != nil {
-				logger.Error(err, "Failed to reconcile Reaper CQL user secret", "ReaperCassandraUserSecretRef", cassandraUserSecretRef)
-				return result.Error(err)
+			if kc.Spec.Reaper != nil && kc.Spec.Reaper.ReaperRef.Name == "" {
```
Comment (rzvoncek) [inline controllers/k8ssandra/secrets.go:66]: @adejanovski , I had to revert this back to the ReaperRef check.
If I checked only a storage type, then a bundled reaper that uses local storage was not working properly.

**Review by adejanovski — APPROVED** [review]

**Review by adejanovski — APPROVED** [review]
